core/src/components/router/utils/matching.ts

```diff
@@ -32,16 +32,58 @@ export const findRouteRedirect = (path: string[], redirects: RouteRedirect[]) =>
   return redirects.find(redirect => matchesRedirect(path, redirect));
 };
 
-export const matchesIDs = (ids: string[], chain: RouteChain): number => {
+export const matchesIDs = (ids: Pick<RouteID, 'id' | 'params'>[], chain: RouteChain): number => {
   const len = Math.min(ids.length, chain.length);
-  let i = 0;
-  for (; i < len; i++) {
-    if (ids[i].toLowerCase() !== chain[i].id) {
+
+  let score = 0;
+
+  for (let i = 0; i < len; i++) {
+    const routeId = ids[i];
+    const routeChain = chain[i];
+    // Skip results where the route id does not match the chain at the same index
+    if (routeId.id.toLowerCase() !== routeChain.id) {
       break;
     }
+    if (routeId.params) {
+      /**
+       * Maps the route's params into a path based on the path variable names,
+       * to compare against the route chain format.
+       *
+       * Before:
+       * ```ts
+       * {
+       *  params: {
+       *    s1: 'a',
+       *    s2: 'b'
+       *  }
+       * }
+       * ```
+       *
+       * After:
+       * ```ts
+       * [':s1',':s2']
+       * ```
+       */
+      const pathWithParams = Object.keys(routeId.params).map(key => `:${key}`);
+      /**
+       * Only compare routes with the chain that have the same number of parameters.
+       */
+      if (pathWithParams.length === routeChain.path.length) {
```

Comment:
Does it make sense to do this check higher up, perhaps at the same time as the `id` check that `break`s out of the loop?

Reply:
+1 to this, it would be nice to only do the `Object.keys().map()` evaluation if we need to.

Reply:
I'm unsure if this check could move outside the bounds of checking `params` existing, since it's dependent on the number of keys on the params object. 

We could however shave a little off the processing time by checking the length prior to mapping the params. It'd be an extremely micro change, but happy to make that tweak 👍

Reply:
Yeah if it's not too hard, could we move the Object.keys check inside of the body of the if?